core/remote/watcher/index.js

```diff
@@ -359,131 +358,133 @@ class RemoteWatcher {
     return squashMoves(doc, was, previousChanges, originalMoves)
   }
 
-  async applyAll(changes /*: Array<RemoteChange> */) /*: Promise<Error[]> */ {
-    const errors = []
-
-    for (let change of changes) {
-      try {
-        await this.apply(change)
-      } catch (err) {
-        log.error({ path: _.get(change, 'doc.path'), err })
-        if (err instanceof MergeMissingParentError) continue
-        errors.push(err)
-      }
-    }
-
-    return errors
+  async applyAll(
+    changes /*: Array<RemoteChange> */,
+    errorsCount /*: number */ = 0
+  ) /*: Promise<Array<{ change: RemoteChange, err: Error }>> */ {
+    const errors = await Promise.mapSeries(
+      changes,
+      async change => await this.apply(change)
+    ).filter(err => err)
+    if (errors.length === errorsCount) return errors
+    return this.applyAll(errors.map(error => error.change), errors.length)
```

Comment:
If we have 2 errors on the first call of `applyError`, and that the 2 errors can be resolved on the second call, we will enter an infinite loop, no? I think we are missing an happy ending path here: `if (errors.length === 0) return []`

Reply:
I don't think so. Here's my thought process:
1. On first call, `errorsCount` is 0 and `errors.length` is 2
2. Since they differ, we call `applyAll` again
3. On second call, `errorsCount` is 2 and `errors.length` is 0
4. Since they differ, we call `applyAll` again
5. On third call, `errorsCount` is 0 and `errors.length` is 0 (we don't have any changes to merge)
6 Since they're equal, line 369 gets us out of the loop

Reply:
OK, it works, but I still think it would be clearer if we didn't have an `errorsCount` parameter and line 369 would be:

```js
if (errors.length === 0 || errors.length === changes.length) return errors
```